blocker/blocker.go

```diff
@@ -198,74 +205,118 @@ func (bl *Blocker) Start() {
 
-	// defer a function that releases the lock
+	// defer a function that updates the database and sets return values
 	defer func() {
-		err := os.Remove(bl.staticNginxCachePurgeLockPath)
-		if err != nil {
-			bl.staticLogger.Errorf("failed to release nginx lock, err %v", err)
-		}
+		bErr := bl.staticDB.MarkAsSucceeded(blocked)
+		fErr := bl.staticDB.MarkAsFailed(failed)
+
+		err = errors.Compose(err, bErr, fErr)
+		succeeded = len(blocked)
+		failures = len(failed)
 	}()
 
-	// open the nginx cache list file
-	f, err := os.OpenFile(bl.staticNginxCachePurgerListPath, os.O_RDWR|os.O_CREATE, 0644)
-	if err != nil {
-		return err
-	}
-	defer func() {
-		e1 := f.Sync()
-		e2 := f.Close()
-		if e1 != nil || e2 != nil {
-			bl.staticLogger.Warnf("Failed to sync and close nginx cache purger list: %s", errors.Compose(e1, e2).Error())
+	for start < len(skylinks) {
+		// check whether we need to escape
+		select {
+		case <-bl.staticCtx.Done():
+			return
+		default:
 		}
-	}()
-	for _, s := range sls {
-		_, err = f.WriteString(s + "\n")
+
+		// calculate the end of the batch range
+		end := start + batchSize
+		if end > len(skylinks) {
+			end = len(skylinks)
+		}
+
+		// grab all skylinks for this batch
+		batch := make([]string, end-start)
+		for i, sl := range skylinks[start:end] {
+			batch[i] = sl.Skylink
+		}
+
+		// send the batch to skyd, if an error occurs and the current batch size
+		// is greater than one, we simply retry with a smaller batch size
+		err = bl.staticSkydAPI.BlockSkylinks(batch)
+
+		// if there's an error, and it's unrelated to the batch we sent, e.g.
+		// connection issue or something, we return here
+		if err != nil && !strings.Contains(err.Error(), unableToUpdateBlocklistErrStr) {
+			return
+		}
+
+		// otherwise if there's an error and the batchsize is larger than 1, we
+		// simply decrease the batch size and continue
+		if err != nil && batchSize > 1 {
+			bl.staticLogger.Tracef("Error occurred while blocking skylinks retrying with batch size %v, err: %v, retrying with smaller batch size...", batchSize, err)
+			batchSize /= blockBatchSizeDivisor
+			continue
+		}
+
+		// if an error occurs add it to the failed array
 		if err != nil {
-			return err
+			if len(batch) == 1 {
+				failed = append(failed, skylinks[start])
+			} else {
+				bl.staticLogger.Errorf("Critical Developer Error, this code should only execute if the length of the batch equals one")
+			}
+		}
+
+		// if no error occurred, add all skylinks from the batch to the
+		// array of blocked skylinks
+		if err == nil {
+			blocked = append(blocked, skylinks[start:end]...)
+		}
+
+		// if no error has occurred yet, update the latest block timestamp
+		if len(failed) == 0 {
+			latest := skylinks[end-1]
+			err = bl.staticDB.SetLatestBlockTimestamp(latest.TimestampAdded)
+			if err != nil && err != database.ErrNoEntriesUpdated {
+				bl.staticLogger.Tracef("blockSkylinks failed to update timestamp: %s", err.Error())
+			}
```

Comment:
This should either happen after all outstanding skylinks are blocked or we should sort the skylinks that need blocking by their timestamps and we should do it in this method. The main concern here is that a server crash and a restart of the process won't omit any skylinks.

Edit: I see that there is sorting done in `SkylinksToBlock` but not in `SkylinksToRetry`. Since we can't know where the list of skyinks to block will come from, the only responsible thing is to sort here.

Reply:
I fixed it differently. Thanks for catching it though, this would have reset it to lower timestamp on success after first failure. Not a huge issue but still it was an oversight. With the retry loop this can actually be removed. Only the caller of `blockSkylinks` needs to care abut latest block timestamp.